core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -261,13 +283,14 @@ class TransactionMarkerChannelManager(config: KafkaConfig,
       }
     }
 
-    val delayedTxnMarker = new DelayedTxnMarker(txnMetadata, appendToLogCallback, txnStateManager.stateReadLock)
+    val delayedTxnMarker = new DelayedTxnMarker(txnMetadata, appendToLogCallback)
```

Comment:
Hmm, appendToLogCallback() calls txnStateManager.getTransactionState(), which also holds a readLock of TransactionStateManager.stateLock. So, to prevent deadlock, we need to address that too?

Reply:
Good point. We can make the full append callback asynchronous.